### PR TITLE
Add script to measure generation benchmarks for different models sizes

### PR DESCRIPTION
**Pull request by KUNAL1612**
**Patch Description**
Added a script that takes in one/many model sizes as input and outputs generation statistics. Addresses #220. Also allows for custom data to be passed as input. 

**Testing steps**
Run this with multiple model sizes to observe WPS counts. Can take multiple models as input and report output for each.

Usage : 

` python generation_benchmarks.py --model_name 2.7b 13b --padding_size 32`

NB: model checkpoint sizes for 66b are huge and the script will require sufficient storage and time tor run.

### CODE REVIEW

File: metaseq/scripts/generation_benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+def download_data(key):
+    links_to_data = {}
+    links_to_data["125m"] = [
+        "https://dl.fbaipublicfiles.com/opt/v1_20220502/125m/reshard-model_part-0.pt",
+        "https://dl.fbaipublicfiles.com/opt/v1_20220502/125m/reshard-model_part-1.pt",
+    ]
+    links_to_data["350m"] = [
+        "https://dl.fbaipublicfiles.com/opt/v1_20220502/350m/reshard.pt"
+    ]
```
Comment (stephenroller) [inline metaseq/scripts/generation_benchmarks.py:172]: Let's drop this one because it's a special snowflake

File: metaseq/scripts/generation_benchmarks.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    with torch.no_grad():
+        model(prompt_ids)[0]
+
+    end = time.time()
+    times_list = torch.tensor(end - begin + prompt_loading).cuda()
+
+    torch.distributed.all_gather(
+        thread_times, times_list, group=dist_utils.get_global_group()
+    )
+    torch.save(thread_times, "./dependencies/times.pt")
```
Comment (stephenroller) [inline metaseq/scripts/generation_benchmarks.py:96]: you don't need to save an external file. you can pass return values up from the distributed.call_main function.

File: metaseq/scripts/generation_benchmarks.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        file = tarfile.open("./dependencies.tar.gz")
+        file.extractall()
+
+    # download model checkpoint
+    for shard in links_to_data[key]:
+        file_name = shard.split("/")[-1]
+        urllib.request.urlretrieve(shard, "./dependencies/" + file_name)
+
+
+def download_prompts():
```
Comment (stephenroller) [inline metaseq/scripts/generation_benchmarks.py:230]: instead of downloading prompts (and cleaning them up), I suggest we just write 10 (short) prompts and keep them as constants at the top of the file.

File: metaseq/scripts/generation_benchmarks.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+def cleanup_prompts(prompt_path):
+    os.remove(prompt_path)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--model_name", nargs="+")
+    parser.add_argument("--prompt_path", nargs="?")
+    parser.add_argument("--padding_size", nargs="?", default="512")
```
Comment (stephenroller) [inline metaseq/scripts/generation_benchmarks.py:249]: can padding be automatically computed from the inputs?

**Review by stephenroller — COMMENTED** [review]
Wow awesome to see this right before my weekend!  Let's try to simplify it and remove the dependencies. @ngoyal2707 may also have suggestions on how we actually implement timing.

We may also wish to switch to using sequence generator so that we can generate N tokens instead of just scoring.

File: metaseq/scripts/generation_benchmarks.py (hunk 3 above)
Comment (KUNAL1612) [reply metaseq/scripts/generation_benchmarks.py:230]: Fair enough for 10 prompts but I was wondering if someone wanted more (the file has up to 1000) then it may just be too much clutter.  Also isn't it generally better to average out numbers from more prompts than from just 10?

File: metaseq/scripts/generation_benchmarks.py (hunk 4 above)
Comment (KUNAL1612) [reply metaseq/scripts/generation_benchmarks.py:249]: Yes as in padding = length of longest sentence. But in the same spirit as having more sentences, having a higher padding size would mean more generated words and hence a more true average?

File: metaseq/scripts/generation_benchmarks.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+        model = models[0]
+
+    model.summon_full_params()
+    model = model.eval()
+
+    begin = time.time()
+    with torch.no_grad():
+        model(prompt_ids)[0]
+
+    end = time.time()
```
Comment (ngoyal2707) [inline metaseq/scripts/generation_benchmarks.py:90]: https://discuss.pytorch.org/t/how-to-measure-time-in-pytorch/26964 is more correct way to measure pytorch / cuda code timing.

your current way doesn't ensure that the work submitted to cuda stream by `model(prompt_ids)[0]` is completed by the time you get to line 90

**Review by ngoyal2707 — CHANGES_REQUESTED** [review]
putting a blocking comment for changes needed for synchronization needed for timing the code

File: metaseq/scripts/generation_benchmarks.py (hunk 5 above)
Comment (KUNAL1612) [reply metaseq/scripts/generation_benchmarks.py:90]: I see. Thanks for the input!

File: metaseq/scripts/generation_benchmarks.py (hunk 3 above)
Comment (stephenroller) [reply metaseq/scripts/generation_benchmarks.py:230]: If all the prompts are the same length, I don't expect any variance at all. Really I think it's okay if we just have 1 prompt, and averaged over many runs of the same prompt.

File: metaseq/scripts/generation_benchmarks.py (hunk 4 above)
Comment (stephenroller) [reply metaseq/scripts/generation_benchmarks.py:249]: In practice, our batching tries to group similarly lengthed examples together to minimize padding. So some padding is realistic, but we actively try to minimize it.

Comment (KUNAL1612) [reply metaseq/scripts/generation_benchmarks.py:249]: hmm I see. So if I have one sentence (as per comment above) and have that run, say, 1000 times, will the padding being equal to the sentence length make sense? In which case padding will no longer be an argument

**Comment by KUNAL1612** [comment]
> putting a blocking comment for changes needed for synchronization needed for timing the code

Addressed these issues

**Review by punitkoura — APPROVED** [review]
Approving since all change requests have been added.

**Review by ngoyal2707 — APPROVED** [review]
didnt review thoroughly but the timing part looks correct to me now
